### PR TITLE
adding handling of nested dict for actions

### PR DESCRIPTION
**Pull request by emarondan**
A possible solution for #187 

- A new function `make_hashable(obj)` was added to convert objects into hashable versions.
- The `actions()` function was updated to use `make_hashable(obj)`.
- A loop was added to iterate over "actions" in `zyte_api_meta`. This allows to convert immutable frozenset objects into mutable dict objects within the "actions" of `zyte_api_meta`.

### CODE REVIEW

File: scrapy_zyte_api/providers.py [hunk 1]
```diff
@@ -120,6 +120,10 @@ async def __call__(  # noqa: C901
                         "e.g. Annotated[Actions, actions([...list of actions...])]."
                     )
                 zyte_api_meta["actions"] = [dict(action) for action in cls.__metadata__[0]]  # type: ignore[attr-defined]
```
Comment (wRAR) [inline scrapy_zyte_api/providers.py:122]: This is the line that makes the reverse conversion, it needs to be updated to be recursive too.

File: scrapy_zyte_api/providers.py [hunk 2]
```diff
@@ -119,7 +119,12 @@ async def __call__(  # noqa: C901
                         "Actions dependencies must be annotated, "
                         "e.g. Annotated[Actions, actions([...list of actions...])]."
                     )
-                zyte_api_meta["actions"] = [dict(action) for action in cls.__metadata__[0]]  # type: ignore[attr-defined]
+                zyte_api_meta["actions"] = [
+                    {
+                        k: dict(v) if isinstance(v, frozenset) else v
```
Comment (wRAR) [inline scrapy_zyte_api/providers.py:124]: Oh, I think I didn't notice the added code is about the reverse conversion :)

I think this doesn't change tuples back to lists? I would expect dicts and lists in the result as it's a wrapper for a JSON API.

**Comment by wRAR** [comment]
This needs a new test, adding a `"selector": {"type": "css", "value": "button#openDescription"}` item to the action in the existing one may be enough.

File: scrapy_zyte_api/providers.py (hunk 2 above)
Comment (emarondan) [reply scrapy_zyte_api/providers.py:124]: The final result here would be a List of dicts. Do we need something different?

Comment (wRAR) [reply scrapy_zyte_api/providers.py:124]: Dict values can be lists, e.g. in the `values` key (looking at `Action` fields, which should match the API).

Comment (emarondan) [reply scrapy_zyte_api/providers.py]: Added a check for tuples on https://github.com/scrapy-plugins/scrapy-zyte-api/pull/188/commits/213182dd77ccfd7df4126040e3979d575db82df6

**Comment by Gallaecio** [comment]
[Test failures are unrelated to this change.](https://github.com/scrapy-plugins/scrapy-zyte-api/issues/190), but the provider job failure is relevant.

**Comment by emarondan** [comment]
Sorry.. forgot to add `# type: ignore[attr-defined]` comment when using `cls.__metadata__[0]` 
Done in https://github.com/scrapy-plugins/scrapy-zyte-api/commit/cba02bb7c1524b74ddf1b1f329c9b4b07ff9190b

**Review by kmike — APPROVED** [review]

**Review by wRAR — APPROVED** [review]
Thanks!

**Pull request merged**